solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/test/java/com/solace/spring/cloud/stream/binder/outbound/JCSMPOutboundMessageHandlerTest.java

```diff
@@ -0,0 +1,188 @@
+		CorrelationData correlationData = new CorrelationData();
+		messageHandler.handleMessage(getMessage(correlationData));
+
+		pubEventHandler.responseReceivedEx(getCorrelationKey());
+
+		correlationData.getFuture().get(100, TimeUnit.MILLISECONDS);
+	}
+
+	@Test()
+	public void test_handleError_withInTimeout() {
+		CorrelationData correlationData = new CorrelationData();
+		Message<String> msg = getMessage(correlationData);
+		messageHandler.handleMessage(msg);
+
+		pubEventHandler.handleErrorEx(createCorrelationKey(correlationData, msg), new JCSMPException("ooooops"), 1111);
+
+		ExecutionException exception = assertThrows(ExecutionException.class, () -> correlationData.getFuture().get(100, TimeUnit.MILLISECONDS));
+		assertEquals("ooooops", exception.getCause().getCause().getMessage());
+	}
+
+	@Test(expected = TimeoutException.class)
+	public void test_responseReceived_withOutTimeout() throws ExecutionException, InterruptedException, TimeoutException {
+		CorrelationData correlationData = new CorrelationData();
+		messageHandler.handleMessage(getMessage(correlationData));
+
+		correlationData.getFuture().get(100, TimeUnit.MILLISECONDS);
+	}
+
+	@Test()
+	public void test_responseReceived_raceCondition() throws ExecutionException, InterruptedException, TimeoutException {
+		CorrelationData correlationDataA = new CorrelationData();
+		messageHandler.handleMessage(getMessage(correlationDataA));
+		CorrelationData correlationDataB = new CorrelationData();
+		messageHandler.handleMessage(getMessage(correlationDataB));
+		CorrelationData correlationDataC = new CorrelationData();
+		messageHandler.handleMessage(getMessage(correlationDataC));
+
+		pubEventHandler.responseReceivedEx(createCorrelationKey(correlationDataB));
+		pubEventHandler.responseReceivedEx(createCorrelationKey(correlationDataA));
+		pubEventHandler.responseReceivedEx(createCorrelationKey(correlationDataC));
+
+		correlationDataA.getFuture().get(100, TimeUnit.MILLISECONDS);
+		correlationDataB.getFuture().get(100, TimeUnit.MILLISECONDS);
+		correlationDataC.getFuture().get(100, TimeUnit.MILLISECONDS);
+	}
+
+	@Test(expected = TimeoutException.class)
+	public void test_responseReceived_messageIdCollision_futureIsOnlyExpectedIfAckReceivedTwice() throws ExecutionException, InterruptedException, TimeoutException {
+		CorrelationData correlationDataA = new CorrelationData();
+		messageHandler.handleMessage(getMessage(correlationDataA));
+
+		pubEventHandler.responseReceivedEx(createCorrelationKey(correlationDataA));
+		correlationDataA.getFuture().get(100, TimeUnit.MILLISECONDS);
+
+
+		CorrelationData correlationDataB = new CorrelationData();
+		messageHandler.handleMessage(getMessage(correlationDataB));
+		correlationDataB.getFuture().get(100, TimeUnit.MILLISECONDS);
+
+		pubEventHandler.responseReceivedEx(createCorrelationKey(correlationDataA));
+		correlationDataB.getFuture().get(100, TimeUnit.MILLISECONDS);
```

Comment:
Are both `correlationDataB.getFuture().get(100, TimeUnit.MILLISECONDS);` expected to throw a `TimeoutException`? If so, you should use `assertThrows(TimeoutException.class() -> correlationDataB.getFuture().get(100, TimeUnit.MILLISECONDS))` instead of `@Test(expected = TimeoutException.class)`.

`@Test(expected = TimeoutException.class)` will only catch the first thrown `TimeoutException` and passes the test without executing the remaining code in the test.

Reply:
Ok fixed